### ISSUE TITLE
Programmatically inject to `<head />`

### ISSUE THREAD

**Issue by JuanM04** [withastro/starlight#474]
### What version of `starlight` are you using?

0.6.1

### What is your idea?

It would be nice to have a way to add tags to `<head />` programmatically. Something like

```js
starlight({
  // CollectionEntry via argument
  head: async ({ slug }) => { 
    return { tag: 'meta', attrs: { ... } }
  }
})
```

### Why is this feature necessary?

This could be useful for opengraph generation:

```js
starlight({
  // CollectionEntry via argument
  head: async ({ slug }) => { 
    return {
      tag: 'meta',
      attrs: { property: 'og:image', content: `${site}/_og/${slug}.png` }
    }
  }
})
```

### Participation

- [X] I am willing to submit a pull request for this issue.

**Comment by delucis** [comment]
Thanks for the issue @JuanM04! We’ll have to think how best to enable this. We’re currently working on supporting component customisation (#415), and that may end up being the recommended way to do this, i.e. you provide a head component that will receive full route data and be able to return what it wants.

There are some technical restrictions that mean we can’t easily use user-provided functions from the config (basically we need to serialize user config to access it in the pages).

For MDX & Markdown content, this could be possible via [a remark plugin that modifies frontmatter](https://docs.astro.build/en/guides/markdown-content/#modifying-frontmatter-programmatically), not sure if that approach is available for Markdoc though.

**Comment by JuanM04** [comment]
> For MDX & Markdown content, this could be possible via [a remark plugin that modifies frontmatter](https://docs.astro.build/en/guides/markdown-content/#modifying-frontmatter-programmatically), not sure if that approach is available for Markdoc though.

This doesn't work as expected. There's a [note about collections and modifying frontmatter](https://docs.astro.build/en/guides/content-collections/#modifying-frontmatter-with-remark), so that method is incompatible with the way Starlight handles frontmatter

**Comment by delucis** [comment]
Right, Starlight would need to do something like spread user-modified frontmatter into entry frontmatter when rendering content here:

https://github.com/withastro/starlight/blob/ed4dd98ec09c3acd2f196e59e9e83979fe9b995c/packages/starlight/index.astro#L12

Would work for [`head` frontmatter](https://starlight.astro.build/reference/frontmatter/#head), but there are frontmatter properties like `sidebar` or `title` which can be used in navigation that would be used _before_ any mutations like that happened.
